ballerina/azure_ai_search_knowledgebase.bal

```diff
@@ -0,0 +1,843 @@
+
+            self.keyFieldName = schemaInfo.keyFieldName;
+            self.vectorFieldNames = schemaInfo.vectorFieldNames.cloneReadOnly();
+            self.allFields = schemaInfo.allFields.cloneReadOnly();
+        }
+    }
+
+    # Ingests documents into the Azure search knowledge base.
+    # + documents - The documents or chunks to ingest (single document, array of documents, or array of chunks)
+    # + return - An `ai:Error` if ingestion fails, otherwise `nil`
+    public isolated function ingest(ai:Chunk[]|ai:Document[]|ai:Document documents) returns ai:Error? {
+        lock {
+            ai:Chunk[]|ai:Error chunks = self.chunk(documents.clone());
+            if chunks is ai:Error {
+                logIfVerboseEnable(self.verbose, string `Failed to chunk documents: ${chunks.message()}}`, chunks);
+                return error ai:Error("Failed to chunk documents before ingestion", chunks);
+            }
+
+            ai:Embedding[]|error embeddings = self.embeddingModel->batchEmbed(chunks);
+            if embeddings is error {
+                logIfVerboseEnable(self.verbose, string `Failed to generate embeddings for documents: ${embeddings.message()}}`, embeddings);
+                return error ai:Error("Failed to generate embeddings for documents", embeddings);
+            }
+            logIfVerboseEnable(self.verbose, string `Generated embeddings for ${embeddings.length().toString()} chunks.`);
+
+            index:IndexDocumentsResult|error uploadResult = self.uploadDocuments(self.indexClient, chunks, self.index, 
+                    embeddings, {[API_KEY_HEADER_NAME]: self.apiKey}, {api\-version: self.apiVersion});
+            if uploadResult is error {
+                logIfVerboseEnable(self.verbose, string `Failed to upload documents to search index: ${uploadResult.message()}}`, uploadResult);
+                return error ai:Error("Failed to upload documents to search index", uploadResult);
+            }
+            
+            // Validate that all documents were successfully indexed
+            foreach index:IndexingResult result in uploadResult.value {
+                if !result.status {
+                    return error ai:Error(string `Failed to index document with key ${result.'key}: ${result.errorMessage ?: "Unknown error"}`);
+                }
+            }
+            
+            return;
+        }
+    }
+
+    # Retrieves relevant chunks for the given query using vector search.
+    #
+    # + query - The text query to search for
+    # + maxLimit - The maximum number of items to return
+    # + filters - Optional metadata filters to apply during retrieval
+    # + return - An array of matching chunks with similarity scores, or an `ai:Error` if retrieval fails
+    public isolated function retrieve(string query, int maxLimit = 10, ai:MetadataFilters? filters = ()) returns ai:QueryMatch[]|ai:Error {
```

Comment:
Methods generally seem to be long. Unless not possible because of `isolated`, would be better to break up into functions. Will also get simplified with things like query expressions.